java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchTreeBasedDictionaryBuilder.java

```diff
@@ -0,0 +1,157 @@
+   */
+  protected int totalValueCount;
+
+  /**
+   * The number of items in the dictionary.
+   */
+  protected int distinctValueCount;
+
+  /**
+   * If null should be encoded.
+   */
+  protected final boolean encodeNull;
+
+  /**
+   * The search tree for storing the value index.
+   */
+  protected TreeSet<Integer> searchTree;
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   * @param encodeNull if null should be encoded.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+    this.comparator.attachVector(dictionary);
+
+    searchTree = new TreeSet<>((index1, index2) -> comparator.compare(index1, index2));
+  }
+
+  public void startBuild() {
+    totalValueCount = 0;
+    distinctValueCount = 0;
+  }
+
+  public void endBuild() {
+    dictionary.setValueCount(distinctValueCount);
+  }
+
+  public V getDictionary() {
+    return dictionary;
+  }
+
+  /**
+   * Try to add all values from the target vector to the dictionary.
+   * @param targetVector the target vector containing values to probe.
+   * @return the number of values actually added to the dictionary.
+   */
+  public int addValues(V targetVector) {
+    int ret = 0;
+    for (int i = 0; i < targetVector.getValueCount(); i++) {
+      if (!encodeNull && targetVector.isNull(i)) {
+        continue;
+      }
+      if (addValue(targetVector, i)) {
+        ret += 1;
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * Try to add an element from the target vector to the dictionary.
+   * @param targetVector the target vector containing new element.
+   * @param targetIndex the index of the new element in the target vector.
+   * @return true if the element is added to the dictionary, and false otherwise.
+   */
+  public boolean addValue(V targetVector, int targetIndex) {
+    totalValueCount += 1;
+
+    // first copy the value to the end of the dictionary
+    dictionary.copyFromSafe(targetIndex, distinctValueCount, targetVector);
+
+    // try to add the value to the dictionary,
+    // if an equal element does not exist.
+    // this operation can be done in O(logn) time.
+    boolean ret = searchTree.add(distinctValueCount);
+    distinctValueCount = searchTree.size();
+    return ret;
+  }
+
+  /**
+   * Gets the sorted dictionary.
+   * Note that given the binary search tree, the sort can finish in O(n).
+   */
+  public void getSortedDictionary(V sortedDictionary) {
```

Comment:
this is probably better named populateSortedDictionary.

Reply:
Good suggestion. Revised accordingly.